### ISSUE TITLE
Scripted Model gave totally wrong result on iOS, but was correct on C++ frontend

### ISSUE THREAD

**Issue by seeker-Liu** [pytorch/pytorch#38186]
## 🐛 Bug
We have a model created, trained and exported by scripting in Python. However, when we are trying to deploy it, it gives totally wrong result on iOS platform. On the other hand, it gives correct result on C++ frontend on my linux devices with exactly same input. After ruling out many other possibilities, we suspect iOS frontend has some bugs triggered by our model.

## To Reproduce
Our workflow is as follow:
1. Train the model in python

2. Scirpt the model with the following code.
```
filename_model = '/path/to/serialized/model'
model = torch.load(filename_model, map_location='cpu').eval()
traced_script_module = torch.jit.script(model)
traced_script_module.save("./exported_model.th")
```

3. Load and use model in our iOS project. (Irrelevant preprocession and postprocession omitted)
Loading:
```
@implementation TorchModule {
 @protected
    torch::jit::script::Module _impl;
}

- (nullable instancetype)initWithFileAtPath:(NSString*)filePath {
  self = [super init];
  if (self) {
    try {
      auto qengines = at::globalContext().supportedQEngines();
      if (std::find(qengines.begin(), qengines.end(), at::QEngine::QNNPACK) != qengines.end()) {
        at::globalContext().setQEngine(at::QEngine::QNNPACK);
      }
      _impl = torch::jit::load(filePath.UTF8String);
      _impl.eval();
    } catch (const std::exception& exception) {
      NSLog(@"%s", exception.what());
      return nil;
    }
  }
  return self;
}
```
Forwarding:
```
@implementation PCMEvaluateTorchModule {
    @private
    torch::Tensor _tempOnset, _tempFrame;
}
```
```
- (void)predict:(void*)buffer frameCount:(int)frameCount frameOutput:(float **)frameOutput onsetOutput:(float **)onsetOutput {
      try {
          
          torch::Tensor tensor = torch::from_blob(b, {1, 1, frameCount, 229});
          torch::autograd::AutoGradMode guard(false);
          torch::AutoNonVariableTypeMode nfn_var_type_mode(true);
          
          auto ans = _impl.forward({tensor}).toTuple();
          int subscript = 0;
          for (const auto &element : ans->elements()) {
              if(subscript == 0) // onset
              {
                  _tempOnset = element.toTensor();
              }
              else if(subscript == 3) // frame
              {
                  _tempFrame = element.toTensor();
              }
              subscript++;
          }
         
          *frameOutput = _tempFrame.data_ptr<float>();
          *onsetOutput = _tempOnset.data_ptr<float>();
         
      } catch (const std::exception& exception) {
          NSLog(@"%s", exception.what());
      }
}

```
Our original model's ```forward()``` method returns 4 tensors, and that becomes a tuple of 4 tensors in C++ interface. 

## Expected behavior
Given that the iOS side code is almost completely same to [official demo](https://github.com/pytorch/ios-demo-app) and is relatively simple. We expect the model will give correct result, which should be very close to the results on Python and C++ frontends.

## Environment

**On Training Machine:** 
Pytorch installed via conda
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 440.82
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.2
[conda] blas                      1.0                         mkl    defaults
[conda] cudatoolkit               10.1.243             h6bb024c_0    defaults
[conda] mkl                       2020.0                      166    defaults
[conda] mkl-service               2.3.0            py38he904b0f_0    defaults
[conda] mkl_fft                   1.0.15           py38ha843d7b_0    defaults
[conda] mkl_random                1.1.0            py38h962f231_0    defaults
[conda] numpy                     1.18.1           py38h4f9e942_0    defaults
[conda] numpy-base                1.18.1           py38hde5b4d6_1    defaults
[conda] pytorch                   1.4.0           py3.8_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.5.0                py38_cu101    pytorch

**On Ubuntu Test Machine:**
OS: Ubuntu 19.10
gcc: 9.2.1
libtorch: 1.4

**On iPhone Test Machine:**
iPhone 8 with iOS 13.
Pytorch 1.4 installed via cocoapods.

For any further environment information needed, please specify.

## Additional context
Our model is not using quantization as for now.
Our model is intending to be used commercially so I am afraid I cannot post the model here, maintainers are welcome to leave your email address and I will send model to you for further investigation.
For any more information needed, please leave comments. 

cc @suo

**Comment by seeker-Liu** [comment]
At the beginning we thought this must be caused by some incorrect settings. But we just can't come up with an idea about where could possibly go wrong.

**Comment by ailzhang** [comment]
Is it possible that you can provide a minimal repro? Thanks!

**Comment by seeker-Liu** [comment]
> Is it possible that you can provide a minimal repro? Thanks!

Yes, regarding the minimal repo, will an iOS minimal project and a scripted model suffice?

**Comment by xta0** [comment]
@seeker-Liu Could you send your model to taox@fb.com? I can help investigate

**Comment by seeker-Liu** [comment]
> @seeker-Liu Could you send your model to [taox@fb.com](mailto:taox@fb.com)? I can help investigate

@xta0 Already sent. Please check your mailbox.

**Comment by xta0** [comment]
@seeker-Liu Thanks, I'll look into it shortly. Can you also try tracing the model instead of scripting? Something like

```
filename_model = '/path/to/serialized/model'
model = torch.load(filename_model, map_location='cpu').eval()
example = some_example_input
traced_script_module = torch.jit.trace(model, example)
traced_script_module.save("./exported_model.th")
```

**Comment by seeker-Liu** [comment]
> @seeker-Liu Thanks, I'll look into it shortly. Can you also try tracing the model instead of scripting? Something like
> 
> ```
> filename_model = '/path/to/serialized/model'
> model = torch.load(filename_model, map_location='cpu').eval()
> example = some_example_input
> traced_script_module = torch.jit.trace(model, example)
> traced_script_module.save("./exported_model.th")
> ```
> 
> let me know if it works or not

@xta0 Tracing won't work because our model contains loops.

**Comment by xta0** [comment]
@seeker-Liu You can separate your loops into another function as mark it as `@torch.jit.script`. Or you can move your for loops to a submodule and script it before tracing.

```
model.my_for_loop_module = torhc.jit.script(model.my_for_loop_module)
scripted_module = torch.jit.trace(model, example)
```
BTW, have you compared the results from your python model?

**Comment by seeker-Liu** [comment]
@xta0 Thanks, I will try.

I have not compared the results from our python model but I can tell it is almost identical to the result of C++ frontend. We already deployed the same model on the server (Ubuntu, C++ frontend) and it works as expected, so the results of C++ and Python must be close.

**Comment by seeker-Liu** [comment]
@xta0 Update: We already reproduced same issue on a simpler model (consists of some conv, activation and dropout layers only).
Do you need that?

**Comment by husthyc** [comment]
Hi @seeker-Liu, I've received your demo project from @xta0 and built & run it. I compared the value of the "ans" (the variable "ans" from your iOS code and the variable "ans" from your C++ code), as well as the value each onset_accessor[i][j] and frame_accessor[0][i][j]. The error of them can be almost ignored, the average error is about 10e-9. It looks like your code has some issue with printing and loading data to/from files.

The error calculation is attached. 

- `result_cxx.txt`: the value of onset_accessor (column 1) and frame_accessor (column 2) from C++
- `result_ios.txt`: the value of onset_accessor (column 1) and frame_accessor (column 2) from iOS 
- You can run `python main.py` to see the result, please make sure your `pandas` has been installed, normally it can be installed by `pip install pandas`

[err_calc.zip](https://github.com/pytorch/pytorch/files/4685906/err_calc.zip)

**Comment by seeker-Liu** [comment]
@husthyc Thanks you, may I ask for the info of your testing environment? And how did you check the value of ```ans``` and print the value of ```onset``` and ```frame```?

**Comment by husthyc** [comment]
@seeker-Liu my test environment is below:
```
Xcode version: 11.4.1
LibTorch version: 1.5.0 (both iOS and macOS)
```

Regarding `ans`,`onset` and `frame`, I did two things:

- Added breakpoints to those places, run the demo, and compared them manually.
- Printed them to `.txt` files as the ones included in the `.zip` file I attached. (in the double for-loop of your code) 
```
for(int frame = 0; frame < fileContentLineNumber; frame++) {
                  for(int row = 0; row < 88; row++) {
...
}
}
```

**Comment by seeker-Liu** [comment]
@husthyc That's it, our productive env is LibTorch 1.4, I will try if upgrade to 1.5 will solve the problem for us.

**Comment by husthyc** [comment]
@seeker-Liu  Tried Libtorch-iOS 1.4.0 just now. Same result.

**Comment by seeker-Liu** [comment]
@husthyc After upgrading, issue persists. Could you recall what device and iOS version you used for testing?

**Comment by husthyc** [comment]
@seeker-Liu I can reproduce the issue on a real device. I was running it on a simulator which gave the same result as C++ front end. Investigating it.

**Comment by seeker-Liu** [comment]
@husthyc Any update?

**Comment by husthyc** [comment]
Hi @seeker-Liu , we did find out some causes. One of them is: aten::conv2d has an optimized version which is "depthwise convolution" on ARM. However, the the output from this function is different from other platforms. We are looking into the details. Do you have any idea if you have any optimized operators in your model?

**Comment by husthyc** [comment]
@seeker-Liu This issue is introduced by https://github.com/pytorch/pytorch/pull/26556. Working on the solution.

**Comment by seeker-Liu** [comment]
> Hi @seeker-Liu , we did find out some causes. One of them is: aten::conv2d has an optimized version which is "depthwise convolution" on ARM. However, the the output from this function is different from other platforms. We are looking into the details. Do you have any idea if you have any optimized operators in your model?

"Optimized operators"? I have no idea about what it is referring to. We have no knowledge about the implementations of ML frameworks. However, I don't think our model contains any.

**Comment by husthyc** [comment]
> > Hi @seeker-Liu , we did find out some causes. One of them is: aten::conv2d has an optimized version which is "depthwise convolution" on ARM. However, the the output from this function is different from other platforms. We are looking into the details. Do you have any idea if you have any optimized operators in your model?
> 
> "Optimized operators"? I have no idea about what it is referring to. We have no knowledge about the implementations of ML frameworks. However, I don't think our model contains any.

Never mind. It's okay. We just want to double check if you did any optimization for some  operators on specific platforms. As I mentioned, the root cause has been found.

**Comment by xta0** [comment]
Hey the fix is on the way - https://github.com/pytorch/pytorch/pull/39591. If you want to try it, please follow the iOS tutorial to build the PyTorch iOS libraries from source code.

**Comment by seeker-Liu** [comment]
> Hey the fix is on the way - #39591.

It is really nice for us to hear that, how long will it takes to appear in nightly-build?

**Comment by yyang9887** [comment]
> 
> 
> Hey the fix is on the way - #39591.

Thank you very much! I have build it for iOS device and got the correct predict result.

**Comment by xta0** [comment]
> > Hey the fix is on the way - #39591.
> 
> It is really nice for us to hear that, how long will it takes to appear in nightly-build?

Should be landed soon. If you can't wait, feel free to manually apply the change and compile from source code

**Comment by xta0** [comment]
The fix has been landed. I'm closing this issue for now, feel free to leave comments below if have any questions.
